### PR TITLE
Gopher Submission 

### PR DESCRIPTION
**Pull request by YuhangP**
This Pull Request is an initial submission for the Gopher game implementation. While the environment is not yet fully executable, the core rendering engine and state structures are established.

### CODE REVIEW

**Comment by BasicTwig** [comment]
Review of Gopher: By the Double Dunk group

ALE-Similarity: 0/4
    > Since the game is not executable, it is impossible to test this category (hence -4 points)
    > Game seems to be not executable because of missing function implementations

Implementation-Quality: 1/4
    Note: At the time of making this review, about a third of the lines of code is actual code, the rest is commented out.
    I will only be reviewing the code that isn't commented out.
    > The _player_step function is missing a return value, despite expecting a "GopherState" value to return 
    > Improper Syntax used in state.replace in line 271-273: replace the state value "new_value_speed" but GopherState doesn't contain a value with that name.
    > Correct usage of the jax and jnp libraries
    > Only basic movement/player step is implemented but there are still errors in the implementation (-3)

**Comment by jakobsteinke** [comment]
Hi, this is the review of the team working on Darkchambers:

### ALE-Similarity: **1 / 4**

- Environment is not executable  
- Major game logic is missing:
  - Gopher behavior
  - Hole/tunnel digging and repair
  - Carrot/seed interactions
  - Collisions
  - Scoring and termination logic

Focus first on implementing missing functions to make the environment executable.

---

### Code Quality: **2 / 4**

#### Bugs found (code crashes)
1. **Undefined state attributes** (Lines 117–136)  
   - `GopherState` attributes are commented out but still referenced (`gopher_position`, `gopher_direction_x`, `gopher_state`, …)  
   - Likely need to uncomment them in the `NamedTuple`

2. **Methods that don’t return anything** (Lines 584–621)  
   - `_handle_tunnel_move()`, `_handle_dig_up()`, `_handle_stealing()`, `_handle_back_tunnel()` only contain `return`  
   - Either implement them properly or remove them from the handler list

3. **Print statements in JIT functions** (Lines 959–960)  
   - Printing inside `@jax.jit render()` is incompatible with JAX  
   - Remove both print statements

4. **Typos / wrong variable names** (Lines 275, 857)  
   - Line 275: `new_player_speed` does not exist (probably `player_speed`)  
   - Line 857: `operand = new_state`, but `new_state` is never defined (likely `state_after_move`)

5. **Missing function arguments** (Lines 584–586)  
   - `jax.lax.cond()` call is missing the `operand` parameter  
   - Add `operand=state`

#### Modularity / Code Organization Issues
1. **Duplicate method definitions**  
   - `_gopher_pop_up()` defined twice (Lines 520 / 719)  
   - Same for `_gopher_steal()` and turn-around methods  
   - Python silently uses the last definition

2. **Dead code blocks (~400 lines)**  
   - Lines 267–448: commented carrot/bonk/tunnel logic (~180 lines)  
   - Lines 453–670: commented gopher AI state machine (~220 lines)  
   - Suggest deleting or moving to a `_legacy.py` reference file

3. **Wrong constant syntax** (Lines 94, 96–99)  
   - `HOLE_POSITION` and `TUNNEL_POSITION` use set syntax `{}` instead of tuples `()`  
   - JAX requires tuples or arrays, not sets

#### JAX-Native / JAX Pattern Recommendations
1. **Sequential conditionals** (Lines 194–229)  
   - Chaining many `jax.lax.cond()` calls is inefficient  
   - Prefer vectorized logic with `jnp.where()`

2. **Python built-ins on JAX arrays** (Line 765)  
   - `len(s.carrots_present)` will not work  
   - Use `s.carrots_present.shape[0]`

3. **Constants rebuilt every frame** (Lines 538, 553, 697)  
   - `jnp.array(HOLE_POSITION)` created repeatedly  
   - Precompute once in `__init__` (e.g. `self.hole_array = jnp.array(...)`)

4. **Unnecessary switching** (Line 882)  
   - `jax.lax.switch()` for 4 states is overkill  
   - Simple `if / elif` is sufficient here

---

### Performance: **0 / 4**

- Environment is not executable  
- Performance cannot be evaluated until core logic is implemented

**Comment by PaulSeitz** [comment]
Hey, overall your implementation looks good, especially the gopher movement and its animations. Also the player animations and movement seem almost perfect. A few notes on things I noticed in the gameplay and the implementation as a whole:

- the animation when the gopher is at the surface and runs into the carotte has a non transparent border
- few graphical things I noticed when comparing the two with the gameplay_comparison.py script:
    - the gopher tunnels are slightly displaced on the y axis (probably about one pixel?)
    - the man is starting at a slightly differnt position
    - slight displacement of the score
- in ALE you cant spam press 'FIRE' to keep digging but have to press-release
- your gopher is pretty aggressive, it sometimes pops up and instantly eats a carrot after starting the game. The ALE version keeps running around and digging for a long time before it starts a run. Might have to tune those parameters a bit
- your game only starts after a timer + pressing fire. The base game just starts after a timer
- and on that note, once you remove the 'press fire' requirement you will notice that the timer timings are different.
- I think I can hit faster with the shovel if I spam click in ALE than I can in your implementation

Implementation wise:
- you have defined a GopherObservation NamedTuple but never use it. Instead you render the frame as the observation, which for our purposes is slightly incorrect as the decision between object centric observations (GopherObservation) and pixel-based observations (the rendered frame) is made in the wrapper layer one level above of the environments. The environments should by default return object-centric observations
- you return a static 'False' for the done flag and a permanent reward of 0.0, if you try to train on this any agent wont learn anything (instead just use the ingame score)
- the code mostly has proper segmentation of concerns, I would just pull the game reset logic from the _gopher_step into its own function. Technically the reset is not part of the gophers_step but its own event (imagine for example you want to modify what happens on a reset in a modification, in your current code you would need to override the entire _gopher_step function even though you only want to change reset behavior)
- performance looks good, the rendering is a bit slow but that might just be due to the larger amount of animations so its fine.

I think that was all I noticed, if I overlooked a bug or something you will probably notice it during agent training later on :)

**Comment by YuhangP** [comment]
> Hey, overall your implementation looks good, especially the gopher movement and its animations. Also the player animations and movement seem almost perfect. A few notes on things I noticed in the gameplay and the implementation as a whole:
> 
>     * ~the animation when the gopher is at the surface and runs into the carotte has a non transparent border~ nevermind, I think that one was my fault
> 
>     * few graphical things I noticed when comparing the two with the gameplay_comparison.py script:
>       
>       * the gopher tunnels are slightly displaced on the y axis (probably about one pixel?)
>       * the man is starting at a slightly differnt position
>       * slight displacement of the score
> 
>     * in ALE you cant spam press 'FIRE' to keep digging but have to press-release
> 
>     * your gopher is pretty aggressive, it sometimes pops up and instantly eats a carrot after starting the game. The ALE version keeps running around and digging for a long time before it starts a run. Might have to tune those parameters a bit
> 
>     * your game only starts after a timer + pressing fire. The base game just starts after a timer
> 
>     * and on that note, once you remove the 'press fire' requirement you will notice that the timer timings are different.
> 
>     * I think I can hit faster with the shovel if I spam click in ALE than I can in your implementation
> 
> 
> Implementation wise:
> 
>     * you have defined a GopherObservation NamedTuple but never use it. Instead you render the frame as the observation, which for our purposes is slightly incorrect as the decision between object centric observations (GopherObservation) and pixel-based observations (the rendered frame) is made in the wrapper layer one level above of the environments. The environments should by default return object-centric observations
> 
>     * you return a static 'False' for the done flag and a permanent reward of 0.0, if you try to train on this any agent wont learn anything (instead just use the ingame score)
> 
>     * the code mostly has proper segmentation of concerns, I would just pull the game reset logic from the _gopher_step into its own function. Technically the reset is not part of the gophers_step but its own event (imagine for example you want to modify what happens on a reset in a modification, in your current code you would need to override the entire _gopher_step function even though you only want to change reset behavior)
> 
>     * performance looks good, the rendering is a bit slow but that might just be due to the larger amount of animations so its fine.
> 
> 
> I think that was all I noticed, if I overlooked a bug or something you will probably notice it during agent training later on :)

>Thank you very much for your feedback; I have already implemented the suggested improvements into my game.

>I had two quick questions regarding the project guidelines. First, the announcement mentions submitting our mods via Moodle, but I am currently unable to locate an active submission link. Second, I’ve been looking for the final report template within the Lab Guideline file as mentioned, but haven't been able to find it.

>Could you please clarify where the submission portal is located and where I might find the template?

>Thank you for your help!

**Comment by PaulSeitz** [comment]
> Good afternoon, @PaulSeitz
> 
> Thank you very much for your feedback. I have already implemented the suggested improvements into my game.
> 
> I had two quick questions regarding the project guidelines. First, the announcement mentions submitting our mods via Moodle, but I am currently unable to locate an active submission link. Second, I’ve been looking for the final report template within the Lab Guideline file as mentioned, but haven't been able to find it.
> 
> Could you please clarify where the submission portal is located and where I might find the template?
> 
> Thank you for your help!

Your ideas what mods you want to implement can just be posted into this PR or sent to any of us per mail. The mods themselves should just be implemented in your branch where your environment is implemented as well, specifically in the mods folder (there you can see how other environments like pong or seaquest already implement mods). 

For the final report there will be no specific template, you can use your preferred way of writing a small document (google doc, latex with the TU template, etc.). The important thing is that you can in the end submit a pdf and have the training metrics of agents trained on your environment + evaluated on your mods. 
I am currently working on an example report which I will try to get published on moodle in the next few days, there I will list / show the specific graphs that we expect. But generally the report itself will not be too much work, as its mostly a collection of graphs you can pull from your wandb training runs.

**Comment by YuhangP** [comment]
Hi,

I have completed the implementation and mods for Gopher environment. This Pull Request is now finalized.
###  What was implemented:
* Full base game implementation in JAX.
* Mods designed for the game.

**5 Simple Mods:**
* **Greedy Gopher:**  Increases the gopher's steal probability (up to 1.0) and decreases peeking time. Tests the agent's ability to adapt to a highly aggressive adversary.
* **Mirror Button:**  Inverts the left and right movement controls. Tests spatial understanding and robustness to action space permutations.
* **Energy Drain:**  Subtracts a constant 0.05 points every frame. Tests if the agent can optimize for speed to avoid bleeding points.
* **Heavy Shovel:**  Increases the action animation to 25 frames, locking movement. Tests if the agent relies on high frequency button spamming or can adapt to longer action commitments.
* **Fast Seed:**  Doubles the falling speed of the seeds. Tests reaction time and the ability to handle accelerated environment dynamics.

**3 Hard Mods:**
* **Invisible Gopher:**  Dynamically hides the gopher while it travels underground. Tests object permanence and the ability to predict trajectories without continuous visual tracking.
* **Wind Gopher:**  Applies a constant wind force pushing the farmer to the left. Tests the agent's ability to compensate for environmental resistance.
* **Dizzy Farmer:**  The gopher's perceived x coordinate wobbles back and forth mathematically. Tests robustness to noisy observations and adversarial hallucinations.

###  To start the game:
```bash
python scripts/play.py -g Gopher
